sklearn/tree/_tree.pyx

```diff
@@ -296,10 +399,17 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
                     is_leaf = (is_leaf or split.pos >= end or
                                (split.improvement + EPSILON <
                                 min_impurity_decrease))
-
-                node_id = tree._add_node(parent, is_left, is_leaf, split_ptr,
-                                         impurity, n_node_samples,
-                                         weighted_n_node_samples, split.missing_go_to_left)
+                with gil:
+                    if parent in self.initial_roots:
+                        node_id = tree._update_node(parent, is_left, is_leaf,
+                                                    split_ptr, impurity, n_node_samples,
+                                                    weighted_n_node_samples,
+                                                    split.missing_go_to_left)
+                    else:
+                        node_id = tree._add_node(parent, is_left, is_leaf,
+                                                 split_ptr, impurity, n_node_samples,
+                                                 weighted_n_node_samples,
+                                                 split.missing_go_to_left)
```

Comment:
This will significantly slow down the tree building because you have to access the GIL again.

If this if parent in self.initial_roots check is necessary, then I think we should just build a C++ hashmap instead, so we don't need the GIL

Reply:
It's an annoying thing as we have node splitting and node updating mixed together. I agree that it's time consuming, and will find a way to simplify it. Which way you think is more applicable, separating the node updates or pushing an indicator onto the stack?

Reply:
Indicator seems reasonable?

Alternatively, why not just make `initial_roots` a hashmap? Using chatgpt, the code seems not too bad:

```
# hashmap_cython.pyx

# Import necessary Cython headers
from libcpp.unordered_map cimport unordered_map

# myclass is the treebuilder
cdef class MyClass:
    cdef unordered_map[int, int] map

    def __cinit__(self):
        self.map = unordered_map[int, int]()

    cpdef int insert(self, int key, int value) nogil:
        # Use NOGIL to allow direct access to the unordered_map
        self.map[key] = value
        return 1

    cpdef int find(self, int key) nogil:
        # Use NOGIL to allow direct access to the unordered_map
        try:
            return self.map.at(key)
        except KeyError:
            return -1
```

Reply:
It seems very complicated to use tuple as hashmap keys, so I separated the node updates and node additions.

Reply:
What is initial roots storing? Hash map of node id as key to what as value?

Reply:
If that is the case, does it makes sense to just refactor it then into:

```
{
     'parent_node_id': {
             'is_left_child': True/False,
             'n_samples': n_samples,
             'depth': depth
     }
}
```

Would this be possible w/o breaking the existing code? I'm just trying to limit code duplication in the fork because otw every time we sync upstream changes, it gets more complex.

Reply:
I'll get back to this soon. Just realized because merging `submodulev2` changed the package name to `scikit-learn-tree`, I was not testing the updated package.

Reply:
No, it defaulted to the `scikit-learn` fork I have before the pr.

Reply:
So you now prefer the original `update` method instead of integrating the function to `build`?

Reply:
Just thinking out diff approaches. If the code as is works now then it's just ideas on how to simplify the implementation so we don't have maintainence work constantly

Reply:
One advantage of the `update` method is that it would not be affected by any upstream changes.

Reply:
True. The issue is that will you be able to account for monotonic constraints and any other added features easily tho?

Reply:
As I already implemented the initial roots method I would stick with it for now. It seems that the errors come from `fit` underperforming, which my code should have no effects on. I'll compare it with the sklearn code tomorrow and see if I can resolve it. It is affecting both the update method and the build/roots method.

Reply:
Let's see if https://github.com/neurodata/scikit-learn/pull/52 will fix it